cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,374 @@
+    ApplicationId applicationId = new ApplicationId(program.getNamespace(), program.getApplication(),
+                                                    program.getVersion());
+    return new ProgramId(applicationId, ProgramType.valueOf(program.getType().toUpperCase()), program.getName());
+  }
+
+  private void deleteAllAppsWithCapability(String capability) throws Exception {
+    doForAllAppsWithCapability(capability, this::deleteAppWithRetry);
+  }
+
+  private void stopAllProgramsWithCapability(String capability) throws Exception {
+    doForAllAppsWithCapability(capability, this::stopAllRunningProgramsForApp);
+  }
+
+  private void deployAllSystemApps(String capability, List<SystemApplication> applications) throws Exception {
+    if (applications.isEmpty()) {
+      LOG.debug("Capability {} do not have apps associated with it", capability);
+      return;
+    }
+    for (SystemApplication application : applications) {
+      doWithRetry(application, this::retryableDeployApp);
+    }
+  }
+
+  private void retryableDeployApp(SystemApplication application) throws Exception {
+    ApplicationId applicationId = getApplicationId(application);
+    LOG.debug("Deploying app {}", applicationId);
+    if (isAppDeployed(applicationId)) {
+      //Already deployed.
+      LOG.debug("Application {} is already deployed", applicationId);
+      return;
+    }
+    String configString = application.getConfig() == null ? null : GSON.toJson(application.getConfig());
+    applicationLifecycleService
+      .deployApp(applicationId.getParent(), applicationId.getApplication(), applicationId.getVersion(),
+                 application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
+  }
+
+  private boolean isAppDeployed(ApplicationId applicationId) throws Exception {
+    try {
+      applicationLifecycleService.getAppDetail(applicationId);
+      return true;
+    } catch (ApplicationNotFoundException exception) {
+      return false;
+    }
+  }
+
+  //Find all applications for capability and call consumer for each
+  private void doForAllAppsWithCapability(String capability, CheckedConsumer<ApplicationId> consumer) throws Exception {
+    for (NamespaceMeta namespaceMeta : namespaceAdmin.list()) {
+      int offset = 0;
+      int limit = 100;
+      NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      EntityResult<ApplicationId> results = getApplications(namespaceId, capability, null,
+                                                            offset, limit);
+      while (!results.getEntities().isEmpty()) {
+        //call consumer for each entity
+        for (ApplicationId entity : results.getEntities()) {
+          consumer.accept(entity);
+        }
+        offset += limit;
+        results = getApplications(namespaceId, capability, results.getCursor(), offset, limit);
+      }
+    }
+  }
+
+  private void stopAllRunningProgramsForApp(ApplicationId applicationId) {
```

Comment:
Just call it `stopAllPrograms(ApplicationId appId)`. It is clear enough what is it doing.

Reply:
Done